csrc/custom_all_reduce.cuh

```diff
@@ -123,47 +131,60 @@ DINLINE O downcast(array_t<float, O::size> val) {
   }
 }
 
-// This function is meant to be used as the first synchronization in the all
-// reduce kernel. Thus, it doesn't need to make any visibility guarantees for
-// prior memory accesses. Note: volatile writes will not be reordered against
-// other volatile writes.
-template <int ngpus>
-DINLINE void start_sync(const RankSignals& sg, volatile Signal* self_sg,
-                        int rank) {
-  if (threadIdx.x < ngpus) {
-    // reset flag for next time
-    self_sg->end[blockIdx.x][threadIdx.x] = 0;
-    // simultaneously write to the corresponding flag of all ranks.
-    // Latency = 1 p2p write
-    sg.signals[threadIdx.x]->start[blockIdx.x][rank] = 1;
-    // wait until we got true from all ranks
-    while (!self_sg->start[blockIdx.x][threadIdx.x]);
-  }
-  __syncthreads();
+static DINLINE void st_flag_release(FlagType* flag_addr, FlagType flag) {
+  asm volatile("st.release.sys.global.u64 [%1], %0;" ::"l"(flag),
+               "l"(flag_addr));
+}
+
+static DINLINE FlagType ld_flag_acquire(FlagType* flag_addr) {
+  FlagType flag;
+  asm volatile("ld.acquire.sys.global.u64 %0, [%1];"
+               : "=l"(flag)
+               : "l"(flag_addr));
+  return flag;
+}
+
+static DINLINE void st_flag_volatile(FlagType* flag_addr, FlagType flag) {
+  asm volatile("st.volatile.global.u64 [%1], %0;" ::"l"(flag), "l"(flag_addr));
+}
+
+static DINLINE FlagType ld_flag_volatile(FlagType* flag_addr) {
+  FlagType flag;
+  asm volatile("ld.volatile.global.u64 %0, [%1];"
+               : "=l"(flag)
+               : "l"(flag_addr));
+  return flag;
```

Comment:
is it possible to avoid directly embedding asm?

Reply:
I don't think so for acquire/release. . Even cuda std c++ lib doesn't have those. For volatile, yes we can use volatile keyword, but casting to volatile is ugly too, so I think stick to PTX is fine.

Reply:
You can use libcu++ for this instead of inline assembly if you feel unhappy about inline assembly: https://nvidia.github.io/cccl/libcudacxx/extended_api/synchronization_primitives/atomic.html#example

The advantage is that you have a distinct type (std::atomic<T> instead of T)